filebeat/tests/integration/delete_utils_test.go

```diff
@@ -0,0 +1,302 @@
+		return nil
+	}
+
+	// Step 5: Parse the response body to extract events
+	var searchResult struct {
+		Hits struct {
+			Hits []struct {
+				Source struct {
+					Message string `json:"message"`
+				} `json:"_source"`
+			} `json:"hits"`
+		} `json:"hits"`
+	}
+	err = json.Unmarshal(body, &searchResult)
+	require.NoError(t, err, "Failed to parse response body")
+
+	// Step 6: Extract the `message` field from each event and return the messages
+	messages := []string{}
+	for _, hit := range searchResult.Hits.Hits {
+		messages = append(messages, hit.Source.Message)
+	}
+
+	return messages
+}
+
+// DisablingProxy is a HTTP proxy that can be disabled/enabled at runtime
+type DisablingProxy struct {
+	mu      sync.RWMutex
+	enabled bool
+	target  *url.URL
+}
+
+// ServeHTTP handles incoming requests and forwards them to the target if
+// the proxy is enabled.
+func (d *DisablingProxy) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	d.mu.RLock()
+	defer d.mu.RUnlock()
+
+	if !d.enabled {
+		http.Error(w, "Proxy is disabled", http.StatusServiceUnavailable)
+		return
+	}
+
+	proxy := httputil.NewSingleHostReverseProxy(d.target)
+	proxy.ServeHTTP(w, r)
+}
+
+// Enable enables the proxy.
+func (d *DisablingProxy) Enable() {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	d.enabled = true
+}
+
+// Disable disables the proxy.
+func (d *DisablingProxy) Disable() {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	d.enabled = false
+}
+
+type FileWatcher struct {
+	watcher       *fsnotify.Watcher
+	targetFile    string
+	mu            sync.Mutex
+	stopChan      chan struct{}
+	eventCallback func(event fsnotify.Event)
+	t             testing.TB
+}
+
+// NewFileWatcher creates a new FileWatcher instance.
+func NewFileWatcher(t testing.TB, targetFile string) *FileWatcher {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		t.Fatalf("failed to create watcher: %s", err)
+	}
+
+	return &FileWatcher{
+		watcher:    watcher,
+		targetFile: targetFile,
+		stopChan:   make(chan struct{}),
+		t:          t,
+	}
+}
+
+// Start begins watching the file's directory for changes.
+func (f *FileWatcher) Start() {
+	dir := filepath.Dir(f.targetFile)
+
+	err := f.watcher.Add(dir)
+	if err != nil {
+		f.t.Fatalf("failed to add directory to watcher: %s", err)
+	}
+
+	go f.watch()
+}
+
+// Stop stops the file-watching process.
+func (f *FileWatcher) Stop() {
+	close(f.stopChan)
+	f.watcher.Close()
+}
+
+// SetEventCallback sets a callback function for file events.
+// To check the event that happened, use event.Has.
+func (f *FileWatcher) SetEventCallback(callback func(event fsnotify.Event)) {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	f.eventCallback = callback
+}
+
+// watch processes file events and errors.
+func (f *FileWatcher) watch() {
+	for {
+		select {
+		case event, ok := <-f.watcher.Events:
+			if !ok {
+				return
+			}
+
+			if event.Name == f.targetFile {
+				f.mu.Lock()
+				if f.eventCallback != nil {
+					f.eventCallback(event)
+				}
+				f.mu.Unlock()
+			}
```

Comment:
I'm not sure this is correct, the location of the mutex let's you change the event callback while handling an event. So if you had a callback for a Remove and that was pending, and then you swapped the callback for on the handled a different event, you'd miss the remove.

Changing the callback should probably only be allowed outside of the context of handling an event.

Reply:
I think I'm over thinking this. There is always a moment in time where you change the callback where the previous event you were subscribed to could be sent next.